HPCC-27152 Helm support for generating remote client certificates for services

afishbeck opened the pull request:
New remote issuer must be enabled (default is not enabled)

```
certificates:
  enabled: true
  issuers:
    remote:
      name: hpcc-remote-issuer
      enabled: true
```

And the issuer requires an additional private key pair
"hpcc-remote-issuer-key-pair", which can be created as follows:

```
openssl req -x509 -newkey rsa:2048 -nodes -keyout remoteca.key -sha256 -days 1825 -out remoteca.crt -config examples/certmanager/ca-req.cfg

kubectl create secret tls hpcc-remote-issuer-key-pair --cert=remoteca.crt --key=remoteca.key

```


Example values config:
---------------------------
```
esp:
- name: eclwatch
  application: eclwatch
  auth: none
  replicas: 1
  remoteClients:
  - name: myclient
```

This will:
1. Make the ESP require client certificates.
2. Create a client certificate for each remoteClient. The certificates and private keys must be set up for the client which varies by client and is outside the scope of this PR.
3. Restricts access to the ESP to listed clients.

Future work may define other ways of adding or using client certificates.

**To use with ecl command line:**

Example ecl.ini entries:

ecl.ini
----------------------
```
eclSSL=true
eclClientCert=/home/jdoe/myclient/tls.crt
eclClientPrivateKey=/home/jdoe/myclient/tls.key
eclCACert=/home/jdoe/myclient/ca.crt
```


example command line parameters:
```
ecl run --ssl --cert=/home/jdoe/myclient/tls.crt --key=/home/jdoe/myclient/tls.key --cacert=/home/jdoe/myclient/ca.crt hthor myquery.ecl
```
Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

afishbeckln committed:
HPCC-27152 Helm support for generating remote client certificates for services

New remote issuer requires an additional private key pair
"hpcc-remote-issuer-key-pair", which can be created as follows:

> openssl req -x509 -newkey rsa:2048 -nodes -keyout remoteca.key -sha256 -days 1825 -out remoteca.crt -config examples/certmanager/ca-req.cfg

> kubectl create secret tls hpcc-remote-issuer-key-pair --cert=remoteca.crt --key=remoteca.key

Example values config:
---------------------------
esp:
- name: eclwatch
  application: eclwatch
  auth: none
  replicas: 1
  remoteClients:
  - name: myclient
    #organization is ptional
    organization: mycompany

For client, example ecl.ini entries:

ecl.ini
----------------------
eclSSL=true
eclClientCert=/home/jdoe/myclient/tls.crt
eclClientPrivateKey=/home/jdoe/myclient/tls.key
eclCACert=/home/jdoe/myclient/ca.crt

example command line parameters:
> ecl run --ssl --cert=/home/jdoe/myclient/tls.crt --key=/home/jdoe/myclient/tls.key --cacert=/home/jdoe/myclient/ca.crt hthor myquery.ecl

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexisrisk.com>

afishbeck commented:
@jakesmith @richardkchapman @ghalliday First draft up for discussion.

Makes it easy to create client certs for a particular service.  We can use them at first for simple mTLS, but later I can generate certificate name based ACLists from these same values.  

i.e. Only allow these specific certificates to work:
myremoteclient42@eclwatch101.eclwatch.hpccnamespace.
myremoteclient42@eclwatch101.eclwatch.hpccnamespace.

richardkchapman reviewed (approved):
Looks sensible to me

ghalliday commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -1337,6 +1337,67 @@ spec:
+*/}}
+{{- define "hpcc.addClientCertificate" }}
+{{- if (.root.Values.certificates | default dict).enabled -}}
+{{- $externalCert := or (and (hasKey . "external") .external) (ne (include "hpcc.isVisibilityPublic" .) "") -}}
```
indent: minor comment, would should the structure more clearly if the ifs were indented - line 1365 is for some reason.

ghalliday reviewed (approved):
One minor comment on indentation.

jakesmith commented on helm/hpcc/templates/dafilesrv.yaml:
```diff
@@ -89,5 +89,12 @@ kind: ConfigMap
+{{- $visibility := .service.visibility -}}
+{{- range $remoteClient := .remoteClients }}
+ {{ include "hpcc.addClientCertificate" (dict "root" $ "client" $remoteClient.name "instance" $instance "component" "dafilesrv" "visibility" $visibility) }}
+{{- end }}
```
the cert. for foreign access may not be appropriate, given that it isn't possible to mix secure/non-secure client access from a foreign environments, and in near future remote won't use certs, but the existing MTLS and signature verificaion route.

jakesmith reviewed (commented):
1 comment.
needs a bit more discussion for specific use case,
but looks good.

jakesmith commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -1337,6 +1337,67 @@ spec:
+  - name: myRemoteClient
+Will generate secrets containing certificates that can be deployed to the remote client.
+If ESP or Roxie is configured to require client certificates then these certificates
+can be used to allow those remote clietns to connect.  Unique certificates should be created
```
typo: "clietns"

jakesmith commented on helm/hpcc/templates/dafilesrv.yaml:
```diff
@@ -89,5 +89,12 @@ kind: ConfigMap
+{{- $instance := .name -}}
+{{- $visibility := .service.visibility -}}
+{{- range $remoteClient := .remoteClients }}
+ {{ include "hpcc.addClientCertificate" (dict "root" $ "client" $remoteClient.name "instance" $instance "component" "dafilesrv" "visibility" $visibility) }}
```
formatting: looks like the helm indentation may be wrong here, would expect perhaps it to be {{- include ..... -}}

afishbeck commented:
@jakesmith updated draft.  I'll convert to non-draft PR soon, but you can take a look, try it out now if you'd like.

1. ESP only for now (roxie needs a bit of work to turn on client cert support and add a cert based ACL.
2. Now supports new explicit remote cert-manager issuer.
3. Generated clients can now be configured with certificates, key, and cacert.
4. Command line works with client certs (although these certs are better for system to system rather than users which will come later).

HPCCSmoketest commented:
Process of PR-15747, label: remoteClientCerts starts now.
The reason of this test is: Forced to re-test.
Commit ID: b2f5e88ce69db85acf3def518fb73314f5cad497
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-29.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Build only for containerized environment.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-29.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: b2f5e88ce69db85acf3def518fb73314f5cad497
Containerized:True
Build: success
Milestone:Install hpccsystems-platform-community_8.6.3-closedown0.el7.x86_64.rpm
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   25 sec (00:00:25) |  421 sec (00:07:01) |  101 sec (00:01:41) |   21 sec (00:00:21) |    0 sec (00:00:00) |    0 sec (00:00:00) |    0 sec (00:00:00) |  568 sec (00:09:28) |

HPCCSmoketest commented:
Process of PR-15747, label: remoteClientCerts starts now.
The reason of this test is: Forced to re-test.
Commit ID: e74ca9f60d318e387ee74230133100bc5bbe3e7e
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-193.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Build only for containerized environment.

afishbeck commented:
@richardkchapman @jakesmith @ghalliday changed from draft to PR.
Now currently ESP only.  Roxie needs a bit of minor refactoring to turn on client cert and ACL support.
Uses a dedicated "remote" certificate issuer.
This now adds support in the generated client code for setting client cert, key, and cacert.
Adds command line support.
- the current client certs are more appropriate certs for individual users.
- We need to think a bit more about user certs if we want to support them)
Adds ecl.ini support for SSL in general and these certs.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-193.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: e74ca9f60d318e387ee74230133100bc5bbe3e7e
Containerized:True
Build: success
Milestone:Install hpccsystems-platform-community_8.6.3-closedown0.el7.x86_64.rpm
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   16 sec (00:00:16) |  397 sec (00:06:37) |  102 sec (00:01:42) |   17 sec (00:00:17) |    0 sec (00:00:00) |    0 sec (00:00:00) |    0 sec (00:00:00) |  532 sec (00:08:52) |

richardkchapman commented on esp/bindings/SOAP/client/soapclient.cpp:
```diff
@@ -145,10 +145,14 @@ int CSoapClient::postRequest(const char* contenttype, const char* soapaction, IR
     {
-        Owned<IHttpClientContext> httpctx = getHttpClientSecretContext(m_mtls_secret);
+        if (m_sec_config)
+          httpctx.setown(createHttpClientContext(LINK(m_sec_config)));
```
Indentation looks dodgy here - tabs?

richardkchapman commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -1370,6 +1376,109 @@ spec:
+  renewBefore: 360h # 15d
+  subject:
+    organizations:
+    - HPCC Systems
```
Is this appropriate, in the OSS helm charts? Not all users will be us...

richardkchapman commented on helm/hpcc/values.yaml:
```diff
@@ -483,6 +495,8 @@ esp:
   application: eclwatch
   auth: none
   replicas: 1
+# remoteClients:
```
Perhaps a better comment would be helpful

richardkchapman reviewed (changes requested):
Some minor comments.

jakesmith commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -1370,6 +1376,109 @@ spec:
 {{- end }}
 
+{{/*
+Builds the commonName for a client certificate.  Used in creation of both certificate and access control list.
```
useful in keeping with [most] other helpers, if comment says what helper template expects to be passed in in dictionary, e.g. "Pass in root, me .. "

jakesmith commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -1370,6 +1376,109 @@ spec:
+  remoteClients:
+  - name: myRemoteClient
+Will generate certificates that can be deployed to the remote client.
+Will cause ESP require client certificates when a socket connects.
```
will cause ESP to require

jakesmith reviewed (commented):
@afishbeck - looks good a few minor comments. I will try to test it.

ghalliday commented on esp/bindings/SOAP/Platform/soapbind.cpp:
```diff
@@ -286,6 +310,8 @@ void CSoapRequestBinding::post(const char *proxy, const char* url, IRpcResponseB
     if (mtls_secret_.length())
         soapclient.setMtlsSecretName(mtls_secret_);
+    if (client_cert_.length() || ca_certs_.length() || accept_self_signed_)
+        soapclient.setSecureSocketConfig(LINK(createSecClientConfig(client_cert_, client_priv_key_, ca_certs_, accept_self_signed_)));
```
Leak: LINK shouldn't be required - because the function returns an object with the link count already incremented.

afishbeck replied:
Fixed.

ghalliday commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -1370,6 +1376,109 @@ spec:
+Builds the commonName for a client certificate.  Used in creation of both certificate and access control list.
+*/}}
+{{- define "hpcc.getClientCommonName" -}}
+  {{- if (.root.Values.certificates | default dict).enabled -}}
```
indent: minor.  I think we agreed on using a single space for indenting helm files.  @jakesmith ?

jakesmith replied:
yes we did.

ghalliday commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -1370,6 +1376,109 @@ spec:
+  - name: myRemoteClient
+Will generate certificates that can be deployed to the remote client.
+Will cause ESP require client certificates when a socket connects.
+Will create a TLS based access contol list which ESP will check to make sure a connections client certificate is enabled.
```
typo: control

ghalliday reviewed (commented):
A few minor comments

afishbeck commented:
Updated based on review:

1. Organization can now be set for client certificate
esp:
  - name: eclwatch
    remoteClients:
    - name: myclient
      organization: mycompany
2. Fixed client code leak
3. Cleaned up indents and comments.

HPCCSmoketest commented:
Process of PR-15747, label: remoteClientCerts starts now.
The reason of this test is: Forced to re-test.
Commit ID: e32358f681a74f952a86cb2268af1d0042169d9f
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-87.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Build only for containerized environment.

jakesmith commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -1370,6 +1376,121 @@ spec:
+{{- end -}}
+
+{{/*
+Turns an array of remoteClients into a | delemited string to be used for the trusted_peers element of SecureSocket settings.
```
typo: delemited -> delimited

jakesmith reviewed (approved):
I've now tested this, have a PR near ready based on this, where DFS uses generated client certificates.
The only caveat (as discussed offline), is that the mechanism requires key and certificate base files, whereas the secret interface returns binary blobs.
As discussed, the workaround for now, will be fetch the local secrets path and use the files directly (that won't work if secrets are in vault though). Th other alternative is to save the secret into local ephemeral storage.

We should explore whether the secure socket implementation can be changed to support keys/certs coming from binary blob to prevent this workaround. Is there JIRA for that?

@afishbeck - 1 typo,  otherwise looks good to squash.

HPCCSmoketest commented:
Process of PR-15747, label: remoteClientCerts starts now.
The reason of this test is: Forced to re-test.
Commit ID: 0b504f68c74c8780ca6d6dc7de73f99bdf73ef30
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-131.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Build only for containerized environment.

afishbeck commented:
Fixed typo, rebased, and created two JIRAs about vault and in memory keys and certificates:

https://track.hpccsystems.com/browse/HPCC-27254
https://track.hpccsystems.com/browse/HPCC-27255

afishbeck commented:
@richardkchapman this should be ready for merging.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-131.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: 0b504f68c74c8780ca6d6dc7de73f99bdf73ef30
Containerized:True
Build: success
Milestone:Install hpccsystems-platform-community_8.6.7-closedown0.el7.x86_64.rpm
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  419 sec (00:06:59) |  101 sec (00:01:41) |   19 sec (00:00:19) |    0 sec (00:00:00) |    0 sec (00:00:00) |    0 sec (00:00:00) |  556 sec (00:09:16) |
